substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/graal/code/SubstrateBackend.java

```diff
@@ -152,14 +166,220 @@ public CompilationResult newCompilationResult(CompilationIdentifier compilationI
+            }
+
+            @Override
+            public int hashCode() {
+                return frame.hashCode();
+            }
+
+            @Override
+            public boolean equals(Object obj) {
+                if (this == obj) {
+                    return true;
+                }
+                if (obj == null || !(obj instanceof BytecodeFrameComparator)) {
+                    return false;
+                }
+                BytecodeFrameComparator that = (BytecodeFrameComparator) obj;
+                BytecodeFrame thisFrame = frame;
+                BytecodeFrame thatFrame = that.frame;
+                while (thisFrame != thatFrame && thisFrame != null && thatFrame != null) {
+                    if (thisFrame.getBCI() == thatFrame.getBCI() &&
+                                    Objects.equals(thisFrame.getMethod(), thatFrame.getMethod()) &&
+                                    Objects.equals(thisFrame.caller(), thatFrame.caller()) &&
+                                    // BytecodeFrame.equals returns true even when these tests fail!
+                                    thisFrame.duringCall == thatFrame.duringCall &&
+                                    thisFrame.rethrowException == thatFrame.rethrowException &&
+                                    thisFrame.numLocals == thatFrame.numLocals &&
+                                    thisFrame.numLocks == thatFrame.numLocks &&
+                                    thisFrame.numStack == thatFrame.numStack &&
+                                    Arrays.equals(thisFrame.values, thatFrame.values)) {
+                        // BytecodeFrame.equals does not check slot kinds!
+                        // The values array may reference a VirtualObject that is
+                        // also in the debuginfo virtual object list. We may need to
+                        // make the compare sensitive to depending on whether the code
+                        // that processes them relies on absolute identity
+                        for (int i = 0; i < thisFrame.numLocals; i++) {
+                            if (thisFrame.getLocalValueKind(i) != thatFrame.getLocalValueKind(i)) {
+                                return false;
+                            }
+                        }
+                        for (int i = 0; i < thisFrame.numStack; i++) {
+                            if (thisFrame.getStackValueKind(i) != thatFrame.getStackValueKind(i)) {
+                                return false;
+                            }
+                        }
+                    } else {
+                        return false;
+                    }
+                    thisFrame = thisFrame.caller();
+                    thatFrame = thatFrame.caller();
+                }
+                return thisFrame == thatFrame;
+            }
+        }
+
+        private static void compressInfoPoints(CompilationResult result) {
+            List<Infopoint> infopointsCopy = new ArrayList<>();
```

Comment:
ArrayList has a java.util.ArrayList#ArrayList(java.util.Collection<? extends E>) constructor that copies the elements.